### PR TITLE
winch: Drop `FuncEnv` trait

### PR DESCRIPTION
**Pull request by saulecabrera**
This commit is a small cleanup to drop the usage of the `FuncEnv` trait.

In https://github.com/bytecodealliance/wasmtime/pull/6358, we agreed on making `winch-codegen` directly depend on `wasmtime-environ`.

Introducing a direct relatioship between `winch-codegen` and `wasmtime-environ` means that the `FuncEnv` trait is no longer serving its original purpose, and we can drop the usage of the trait and use the types exposed from `winch-codegen` directly instead.

Even though this change drops the `FuncEnv` trait, it still keeps a `FuncEnv` struct, which is used during code generation.

### CODE REVIEW

File: winch/codegen/src/codegen/mod.rs [hunk 1]
```diff
@@ -18,50 +18,47 @@ pub use env::*;
 pub mod call;
 
 /// The code generation abstraction.
-pub(crate) struct CodeGen<'a, A, M>
+pub(crate) struct CodeGen<'a, A, M, P>
 where
     M: MacroAssembler,
     A: ABI,
+    P: PtrSize,
```
Comment (alexcrichton) [inline winch/codegen/src/codegen/mod.rs:25]: I'm not all that familiar with the design of these generics, but I would naively expect that a choice of either `MacroAssembler` or `ABI` would imply `PtrSize`. If that's the case would it perhaps be possible to have `P` be an associated type of either prior trait?

Or would it, for example, make sense to put both `ABI` and `PtrSize` inside of the `MacroAssembler` trait as an associated type?

**Review by alexcrichton — APPROVED** [review]

File: winch/codegen/src/codegen/mod.rs (hunk 1 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/mod.rs:25]: That makes sense. Taking a closer look, I'm leaning towards having the `PtrSize` as an associated type inside the `ABI` trait, which already provides similar information (e.g. `word_bytes`).

Comment (saulecabrera) [reply winch/codegen/src/codegen/mod.rs:25]: Upon a second look, having both of these as associated types in the `MacroAssembler` makes more sense I think, and it's going to result in quite a decent cut in boilerplate. I started working on the change, but I'm anticipating that it's going to be a decent amount of changes unrelated to this PR, so I think I prefer landing this and putting the refactor once this lands.

Comment (saulecabrera) [reply winch/codegen/src/codegen/mod.rs:25]: It's also in my todo to standardize the signatures in the `ABI` trait (e.g. there's no reason to have them borrow `&self`), so I'm thinking I'll bundle those improvements in the refactoring too.

**Pull request merged**